### PR TITLE
2674 e2e improve setup sac

### PR DESCRIPTION
**Pull request by sambodeme**
## Description
This PR introduces a series of refactoring changes around the setup_sac logic:
- Updated the names of various functions, variables, and named tuples for improved clarity and consistency.
- Removed sections of code that were no longer in use to streamline the codebase.
- Segmented the existing code into separate files. This enhances readability and maintainability by organizing related functionalities more logically.
- Reorganized the structure of the existing code to improve flow and comprehension.
- Ticket is #2674 

For Testing:

- To generate the necessary workbooks, run the following command:
```docker compose run --rm web python manage.py historic_workbook_generator --year 22 --output data --dbkey 171944```
  This command will create a folder named data inside your project and populate it with the generated workbooks.

- To create and process an audit report, use this command:
```docker compose run --rm web python manage.py historic_data_migrator --years 22 --email your_email_here --dbkeys 182926```
This process will create an audit report, submit it, and then disseminate it. Make sure that you have logged into the application with the specified email address before running this command. 

## PR checklist: submitters

- [ ]   Link to an issue if possible. If there’s no issue, describe what your branch does. Even if there is an issue, a brief description in the PR is still useful.
- [ ]   List any special steps reviewers have to follow to test the PR. For example, adding a local environment variable, creating a local test file, etc.
- [ ]   For extra credit, submit a screen recording like [this one](https://github.com/GSA-TTS/FAC/pull/1821).
- [ ]   Make sure you’ve merged `main` into your branch shortly before creating the PR. (You should also be merging `main` into your branch regularly during development.)
- [ ]   Make sure you’ve accounted for any migrations. When you’re about to create the PR, bring up the application locally and then run `git status | grep migrations`. If there are any results, you probably need to add them to the branch for the PR. Your PR should have only **one** new migration file for each of the component apps, except in rare circumstances; you may need to delete some and re-run `python manage.py makemigrations` to reduce the number to one. (Also, unless in exceptional circumstances, your PR should not delete any migration files.)
- [ ]   Make sure that whatever feature you’re adding has tests that cover the feature. This includes test coverage to make sure that the previous workflow still works, if applicable.
- [ ]   Make sure the full-submission.cy.js [Cypress test](https://github.com/GSA-TTS/FAC/blob/main/docs/testing.md#end-to-end-testing) passes, if applicable.
- [ ]   Do manual testing locally. Our tests are not good enough yet to allow us to skip this step. If that’s not applicable for some reason, check this box.
- [ ]   Verify that no Git surgery was necessary, or, if it was necessary at any point, repeat the testing after it’s finished.
- [ ]   Once a PR is merged, keep an eye on it until it’s deployed to dev, and do enough testing on dev to verify that it deployed successfully, the feature works as expected, and the happy path for the broad feature area (such as submission) still works.

## PR checklist: reviewers

- [ ]   Pull the branch to your local environment and run `make docker-clean; make docker-first-run && docker compose up`; then run `docker compose exec web /bin/bash -c "python manage.py test"`
- [ ]   Manually test out the changes locally, or check this box to verify that it wasn’t applicable in this case.
- [ ]   Check that the PR has appropriate tests. Look out for changes in HTML/JS/JSON Schema logic that may need to be captured in Python tests even though the logic isn’t in Python.
- [ ]   Verify that no Git surgery is necessary at any point (such as during a merge party), or, if it was, repeat the testing after it’s finished.

The larger the PR, the stricter we should be about these points.

### CODE REVIEW

File: backend/census_historical_migration/base_field_maps.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from collections import namedtuple as NT
+
+FormFieldInDissem = WorkbookFieldInDissem = 1000
+
+# This provides a way to map the sheet in the workbook to the
+# column in the DB. It also has a default value and
+# the type of value, so that things can be set correctly
+# before filling in the XLSX workbooks.
+# Define a new named tuple with an additional field
+SheetFieldMap = NT("SheetFieldMap", "in_sheet in_db in_dissem default type")
```
Comment (jadudm) [inline backend/census_historical_migration/base_field_maps.py:10]: Strictly as a comment, not a proposed revision:

These are currently named tuples, or "records." If turning these into classes would make them more useful, it is completely reasonable to do so. FYI.

File: backend/census_historical_migration/sac_general_lib/audit_information.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    CensusCfda22 as Cfda,
+    CensusFindings22 as Finding,
+)
+from census_historical_migration.base_field_maps import FormFieldMap, FormFieldInDissem
+from census_historical_migration.sac_general_lib.utils import (
+    _create_json_from_db_object,
+)
+import audit.validators
+
+DOLLAR_THRESHOLD = 750000
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/audit_information.py:12]: It would be completely reasonable to move this out to the app-level `config`, for a variety of reasons. Perhaps not part of this commit, but we are going to need this number elsewhere.

File: backend/census_historical_migration/sac_general_lib/audit_information.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+mappings = [
+    FormFieldMap(
+        "dollar_threshold", None, FormFieldInDissem, DOLLAR_THRESHOLD, int
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/audit_information.py:16]: There is. 

The draft dictionary (which needs to get published...)

https://federalist-35af9df5-a894-4ae9-aa3d-f6d95427c7bc.sites.pages.cloud.gov/preview/gsa-tts/fac-transition-site/jadudm/dictionaries/developers/v1_0_2/dictionary/

may help?

File: backend/census_historical_migration/sac_general_lib/audit_information.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        "reportablecondition",
+        FormFieldInDissem,
+        None,
+        bool,
+    ),
+    FormFieldMap(
+        "is_low_risk_auditee", None, FormFieldInDissem, False, bool
+    ),  # FIXME: There is no in_db mapping?
+    FormFieldMap(
+        "agencies", None, "agencies_with_prior_findings", [], list
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/audit_information.py:56]: This needs to be fixed in the dictionary. I think this is `PYSCHEDULE`. Maybe.

File: backend/census_historical_migration/sac_general_lib/audit_information.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    ),  # FIXME: There is no in_db mapping?
+]
+
+
+def _get_agency_prefixes(dbkey):
+    agencies = {}
+    cfdas = Cfda.select().where(Cfda.dbkey == dbkey)
+    cfda: Cfda
+    for cfda in cfdas:
+        agencies[int((cfda.cfda).split(".")[0])] = 1
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/audit_information.py:66]: It might be this can be a bit cleaner, and perhaps use a `set` instead of the weird use of a hash table to simulate a set. Or, you can leave it. I know I wrote that code, and perhaps it can be clearer as to its purpose. Up to you.

File: backend/census_historical_migration/sac_general_lib/audit_information.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    return agencies.keys()
+
+
+def _get_gaap_results(dbkey):
+    findings = Finding.select().where(Finding.dbkey == dbkey)
+    gaap_results = {}
+    # FIXME: How do we retrieve gaap_results from the historic data? I could not find corresponding fields in Census tables.
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/audit_information.py:73]: It has a different name. We might ask @Carley-Jackson .

File: backend/census_historical_migration/sac_general_lib/auditee_certification.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+#     FormFieldMap("is_complete_and_accurate", None, FormFieldInDissem, False, bool),
+#     FormFieldMap("has_engaged_auditor", None, FormFieldInDissem, False, bool),
+#     FormFieldMap("is_issued_and_signed", None, FormFieldInDissem, False, bool),
+#     FormFieldMap("is_FAC_releasable", None, FormFieldInDissem, False, bool),
+# ]
+
+auditee_signature_mappings = [
+    FormFieldMap("auditee_name", "auditeename", FormFieldInDissem, None, str),
+    FormFieldMap("auditee_title", "auditeetitle", FormFieldInDissem, None, str),
+    # FIXME: Do we need this ?
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/auditee_certification.py:25]: I don't think we disseminate it. So, possibly/probably not. I'm working from memory, though.

File: backend/census_historical_migration/sac_general_lib/auditee_certification.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+def _auditee_certification(dbkey):
+    gobj: Gen = Gen.select().where(Gen.dbkey == dbkey).first()
+    certification = {}
+    # certification["auditee_certification"] = _create_json_from_db_object(
+    #     gobj, auditee_certification_mappings
+    # )
+    certification["auditee_signature"] = _create_json_from_db_object(
+        gobj, auditee_signature_mappings
+    )
+    # FIXME: Do we need this ?
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/auditee_certification.py:41]: When I hacked this in, I added *just enough data* to pass the cross-validations. I'm... on one hand, it might be nice to be complete, but I don't think we have the historic data to populate the checkboxes. Or: _by definition_, they had to accept the terms in order to submit the FAC. So, _by definition_, our reconstruction would just "automatically" check all the boxes.

Use your judgement here. This is an object that is far, far lower in the integrity chain than, say, `federal_awards`. If the critical information (signature, role, etc.) are present, then building the full object with the booleans for the checkmarks is probably not needed.

If I understand the "do we need this?" question correctly. Happy to discuss.

File: backend/census_historical_migration/sac_general_lib/auditor_certification.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
+# import audit.validators
+from census_historical_migration.workbooklib.census_models.census import (
+    CensusGen22 as Gen,
+)
+from census_historical_migration.base_field_maps import FormFieldMap, FormFieldInDissem
+from census_historical_migration.sac_general_lib.utils import (
+    _create_json_from_db_object,
+)
+
+# FIXME: Do we need this ?
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/auditor_certification.py:10]: See previous comment. Probably not.

File: backend/census_historical_migration/sac_general_lib/general_information.py
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/general_information.py:1]: An overall comment: the restructuring of the forms (having not even finished reading through) is wonderful. Thank you.

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    FormFieldMap("auditor_address_line_1", "cpastreet1", FormFieldInDissem, None, str),
+    FormFieldMap("auditor_city", "cpacity", FormFieldInDissem, None, str),
+    FormFieldMap("auditor_contact_name", "cpacontact", FormFieldInDissem, None, str),
+    FormFieldMap("auditor_contact_title", "cpatitle", FormFieldInDissem, None, str),
+    FormFieldMap("auditor_country", "cpacountry", FormFieldInDissem, None, str),
+    FormFieldMap("auditor_ein", "auditor_ein", FormFieldInDissem, None, str),
+    FormFieldMap("auditor_ein_not_an_ssn_attestation", None, None, True, bool),
+    FormFieldMap(
+        "auditor_email", "cpaemail", FormFieldInDissem, "noemailfound@noemail.com", str
+    ),  # FIXME: Do we want to keep `noemailfound@noemail.com`?
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/general_information.py:43]: No. We should pick something, though.

`fac-no-email-in-census-data@gsa.gov` or similar.

Whatever we use, if we use it, should be descriptive as to why it is being inserted.

Comment (jadudm) [reply backend/census_historical_migration/sac_general_lib/general_information.py:43]: I think our process would actually want to treat this as a failure in the first instance? This might be where we want a mapping function.

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+)
+from census_historical_migration.base_field_maps import FormFieldMap, FormFieldInDissem
+from census_historical_migration.sac_general_lib.utils import (
+    _create_json_from_db_object,
+    _boolean_field,
+)
+
+mappings = [
+    FormFieldMap(
+        "auditee_fiscal_period_start", "fyenddate", "fy_start_date", None, str
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/general_information.py:17]: It is possible (perhaps not in this commit, but soon) that we want an xform that replaces all foundational Python types.

```
def string_to_string(...):
    # error checking, normalization
    # throw an exception if some condition happens
    return str(input)
```

or similar.

This way, we have a centralized xform for all strings, all bools, etc.

I'm particularly thinking about the email question, below. It might want its own special transform, to handle the missing case. (Or, perhaps the missing case/exception/error is part of a standard string-to-string transform).

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    ),  # FIXME: There is no in_db mapping ?
+    FormFieldMap("secondary_auditors_exist", "multiple_cpas", None, None, bool),
+]
+
+
+def _period_covered(s):
+    return {"A": "annual", "B": "biennial", "O": "other"}[s]
+
+
+def _census_audit_type(s):
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/general_information.py:67]: I'm assuming a future version of this will incorporate some kind of error handling for when things are not present in the dictionary?

Again, flagging, but not for this PR. I think the cleanup is excellent, and gives us a staging ground for next steps.

File: backend/census_historical_migration/sac_general_lib/sac_creator.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        user=user,
+        email=user.email,
+        role="editor",
+    )
+
+    # We need these to be different.
+    Access.objects.create(
+        sac=sac,
+        user=user,
+        email="bob_the_auditee_official@auditee.org",  # user.email, FIXME - Confirm this is the right value?
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/sac_creator.py:59]: Like other values, we might need this to be more descriptive? Not sure what we should do here.

Actually, I do.

For now, I think we clean it up a bit. However, this needs a ticket and some discussion.

We're going to either want to insert a clearly defined placeholder, or we're going to want to pull values from the General form for the auditor and auditee contact, but... perhaps label them as choices we made.

What I'm concerned about is that these are _access objects_, and therefore they'll end up in our system. We're creating them because they get accessed as part of the migration. 

Hence... using one set of "fake" access objects makes sense.

But, they should still be `fac-census-migration-auditee-official` or similar.

File: backend/census_historical_migration/sac_general_lib/sac_creator.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    Access.objects.create(
+        sac=sac,
+        user=user,
+        email="bob_the_auditor_official@auditor.org",  # user.email, FIXME - Confirm this is the right value?
+        role="certifying_auditor_contact",
+    )
+
+    sac.auditee_certification = _auditee_certification(dbkey)
+    sac.auditor_certification = _auditor_certification(dbkey)
+    sac.data_source = "CENSUS"  # FIXME - Confirm this is the right value?
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/sac_creator.py:71]: We're using `GSAFAC` and `CENSUS` as keys, so yes. If these want to be broken out to a "constant" somewhere, that's fine.

File: backend/census_historical_migration/workbooklib/end_to_end_core.py [hunk 15]
```diff
@@ -191,19 +191,19 @@ def api_check(json_test_tables):
...
 
 
-def generate_workbooks(user, email, dbkey, year):
+def generate_workbooks(user, dbkey, year):
     entity_id = "DBKEY {dbkey} {year} {date:%Y_%m_%d_%H_%M_%S}".format(
         dbkey=dbkey, year=year, date=datetime.now()
     )
     sac = setup_sac(user, entity_id, dbkey)
     if sac.general_information["audit_type"] == "alternative-compliance-engagement":
         print(f"Skipping ACE audit: {dbkey}")
```
Comment (jadudm) [inline backend/census_historical_migration/workbooklib/end_to_end_core.py:200]: Definitely could become a log statement, and perhaps should become a logged "failure to migrate" when we see ACEE audits (for now).

File: backend/census_historical_migration/workbooklib/excel_creation_utils.py [hunk 16]
```diff
@@ -62,7 +57,7 @@ def set_uei(Gen, wb, dbkey):
     if g.uei:
         set_range(wb, "auditee_uei", [g.uei])
     else:
-        g.uei = "BADBADBADBAD"
+        g.uei = "BADBADBADBAD"  # FIXME: Shouldn't this raise an exception here?
```
Comment (jadudm) [inline backend/census_historical_migration/workbooklib/excel_creation_utils.py:60]: Now it should, yes. Although, it will be 100% true for all audits before 2022. So, we have to have a strategy.

We can discuss. It should not be `BADBADBADBAD`, as you suspect.

File: backend/census_historical_migration/workbooklib/excel_creation_utils.py [hunk 17]
```diff
@@ -92,43 +87,32 @@ def map_simple_columns(wb, mappings, values):
...
-    dt = _census_date_to_datetime(g.fyenddate)
-    return f"{g.audityear}-{dt.month:02}-TSTDAT-{dbkey.zfill(10)}"
+def add_hyphen_to_zip(zip):
+    strzip = str(zip)
+    if len(strzip) == 5:
+        return strzip
+    elif len(strzip) == 9:
+        return f"{strzip[0:5]}-{strzip[5:9]}"
+    else:
+        logger.info("ZIP IS MALFORMED IN WORKBOOKS E2E / SAC_CREATION")
```
Comment (jadudm) [inline backend/census_historical_migration/workbooklib/excel_creation_utils.py:97]: Possibly another place that should fail? Dunno.

File: backend/census_historical_migration/workbooklib/post_upload_utils.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        # nothing to do here
+        return
+
+    with open(pdf_filename, "rb") as f:
+        content = f.read()
+    file = SimpleUploadedFile(pdf_filename, content, "application/pdf")
+    print(file.__dict__)
+    pdf_file = PDFFile(
+        file=file,
+        component_page_numbers={  # FIXME - How do we want to handle these values?
```
Comment (jadudm) [inline backend/census_historical_migration/workbooklib/post_upload_utils.py:90]: They might actually exist somewhere in the historic data. For now, I don't know. We should start a ticket of questions like this. Or something.

File: backend/census_historical_migration/base_field_maps.py (hunk 1 above)
Comment (sambodeme) [reply backend/census_historical_migration/base_field_maps.py:10]: Agreed, actually I took that approach in my initial commit but I can remember why I diverted path.

File: backend/census_historical_migration/sac_general_lib/audit_information.py (hunk 3 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/audit_information.py:16]: Thanks for sharing this. I was struggling searching in googlesheets

File: backend/census_historical_migration/sac_general_lib/auditee_certification.py (hunk 7 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/auditee_certification.py:25]: We don't disseminate, I simply added this comment so that we can make a group decision.

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+    ),
+    FormFieldMap(
+        "user_provided_organization_type", None, "entity_type", "unknown", str
+    ),  # FIXME: There is no in_db mapping ?
+    FormFieldMap("secondary_auditors_exist", "multiple_cpas", None, None, bool),
+]
+
+
+def _period_covered(s):
+    period_dict = {"A": "annual", "B": "biennial", "O": "other"}
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/sac_general_lib/general_information.py:64]: Could this be a const defined somewhere so it's not recreated every time this is called?

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+        raise KeyError(f"Key '{s}' not found in period coverage mapping")
+    return period_dict[s]
+
+
+def _census_audit_type(s):
+    audit_type_dict = {
+        "S": "single-audit",
+        "P": "program-specific",
+        "A": "alternative-compliance-engagement",
+    }
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/sac_general_lib/general_information.py:75]: Similar here, could we define a const outside of the method?

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+    ),  # FIXME: There is no in_db mapping ?
+    FormFieldMap("secondary_auditors_exist", "multiple_cpas", None, None, bool),
+]
+
+
+def _period_covered(s):
+    period_dict = {"A": "annual", "B": "biennial", "O": "other"}
+    if s not in period_dict:
+        raise KeyError(f"Key '{s}' not found in period coverage mapping")
+    return period_dict[s]
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/sac_general_lib/general_information.py:67]: I feel a `try... except` block would be more straightforward here

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+
+def _census_audit_type(s):
+    audit_type_dict = {
+        "S": "single-audit",
+        "P": "program-specific",
+        "A": "alternative-compliance-engagement",
+    }
+    if s not in audit_type_dict:
+        raise KeyError(f"Key '{s}' not found in census audit type mapping")
+    return audit_type_dict[s]
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/sac_general_lib/general_information.py:78]: Similar here, `try... except`?

File: backend/census_historical_migration/sac_general_lib/sac_creator.py [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+    sac.save()
+
+    logger.info("Created single audit checklist %s", sac)
+    return sac
+
+
+def setup_sac(user, test_name, dbkey):
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/sac_general_lib/sac_creator.py:78]: While we're on this ticket I think it's a good time to put a doc string on this

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 11 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:17]: I created a [ticket](https://github.com/GSA-TTS/FAC/issues/2833) for this to limit the size of the PR

File: backend/census_historical_migration/sac_general_lib/auditee_certification.py (hunk 8 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/auditee_certification.py:41]: I opted to build the entire object so that we can utilize the validators.

File: backend/census_historical_migration/sac_general_lib/auditor_certification.py (hunk 9 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/auditor_certification.py:10]: Similar to my previous comment

File: backend/census_historical_migration/sac_general_lib/auditee_certification.py (hunk 7 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/auditee_certification.py:25]: I decided to keep the field so that we don't run into issue with the validator.

File: backend/census_historical_migration/sac_general_lib/audit_information.py (hunk 6 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/audit_information.py:73]: [Ticket ](https://github.com/GSA-TTS/FAC/issues/2832) created for follow up.

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 10 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:43]: We should treat this as a failure

File: backend/census_historical_migration/sac_general_lib/sac_creator.py (hunk 13 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/sac_creator.py:59]: Using `fac-census-migration-auditee-official@auditee.org` for now but also added this to the [ticket](https://github.com/GSA-TTS/FAC/issues/2832)

File: backend/census_historical_migration/workbooklib/post_upload_utils.py (hunk 18 above)
Comment (sambodeme) [reply backend/census_historical_migration/workbooklib/post_upload_utils.py:90]: https://github.com/GSA-TTS/FAC/issues/2832

File: backend/census_historical_migration/workbooklib/excel_creation_utils.py (hunk 17 above)
Comment (sambodeme) [reply backend/census_historical_migration/workbooklib/excel_creation_utils.py:97]: json-schema should catch this.

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 21 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:67]: We want to raise an exception here to fail the migration process

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 22 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:78]: Similar to my previous comment

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 11 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:17]: Actually, I saw a need for it and implemented this method and few others as part of the current PR. The ticket mentioned above will be use to improve that `string_to_string` method and add others (if necessary)

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 19 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:64]: This dictionary is small and unlikely to expand, and since it's only called once per submission (unlike other sections), keeping it within the function is better for readability and maintenance. However, I'll move it out due to its values being duplicated in `Base.libsonnet`, aligning with other code areas where we've adopted a similar approach for duplicated sources. This will help consolidate it in future.

File: backend/census_historical_migration/sac_general_lib/general_information.py (hunk 20 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:75]: Similar to my comment above.

**Review by danswick — APPROVED** [review]
Discussed with @sambodeme 👍

File: backend/census_historical_migration/workbooklib/excel_creation_utils.py (hunk 17 above)
Comment (jadudm) [reply backend/census_historical_migration/workbooklib/excel_creation_utils.py:97]: For the future/future work: yes, but I think *all* Census zips lack the hyphen. So, we may end up wanting a xform that takes 9-digit zips and adds the hyphen, or lots of data won't migrate. In this case, it isn't that the data is "bad," it's just we might be expecting a different form. (Totally may be mis-remembering.)

I can't remember what we expect. I know this merged, but noting it here @sambodeme, @phildominguez-gsa .

Comment (sambodeme) [reply backend/census_historical_migration/workbooklib/excel_creation_utils.py:97]: It looks like what you're describing, `... that takes 9-digit zips and adds the hyphen,` is what line 95 achieves. So either you meant to pull this into a separate xform for consistency, or you meant something else. I will ticket this and follow up with you.
